### PR TITLE
Adding map type write support

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java [hunk 1]
```diff
@@ -419,7 +439,37 @@ private static Object convertSparkValueToProtoRowValue(
     }
 
     if (sparkType instanceof MapType) {
-      throw new IllegalArgumentException(MAPTYPE_ERROR_MESSAGE);
+      // converting the map into ARRAY<STRUCT<key, value>> and making a recursive call
+      MapType mapType = (MapType) sparkType;
+      // As the same map<key, value> will be converted to the same STRUCT<key, value>, it is best to
+      // cache this conversion instead of re-creating it for every value
+      StructType mapStructType = MAP_TYPE_STRUCT_TYPE_CACHE.getUnchecked(mapType);
+      scala.collection.Map map = (scala.collection.Map) sparkValue;
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java:447]: @sonatype-lift ignore

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java [hunk 2]
```diff
@@ -72,7 +81,18 @@ public class ProtobufUtils {
   // For every message, a nested type is name "STRUCT"+i, where i is the
   // number of the corresponding field that is of this type in the containing message.
   private static final String RESERVED_NESTED_TYPE_NAME = "STRUCT";
-  private static final String MAPTYPE_ERROR_MESSAGE = "MapType is unsupported.";
+
+  private static final LoadingCache<MapType, StructType> MAP_TYPE_STRUCT_TYPE_CACHE =
+      CacheBuilder.<MapType, StructType>newBuilder()
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java:86]: @sonatype-lift ignore

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java:447]: @sonatype-lift ignore

**Review by vishalkarve15 — APPROVED** [review]
LGTM

**Pull request merged**